### PR TITLE
Aggregation Logic

### PR DESCRIPTION
**Pull request by Reisen**
This PR includes only commits required to merge aggregation logic into the Pyth contract, these have been spliced out from #139 with only the required utilities for aggregation. This doesn't include utilities that aren't used as they can be included in future PRs.

### CODE REVIEW

**Comment by ali-bahjati** [comment]
Shall we update the tests before review?

`pyth/tests/qset` tests are failing, those are ran with `pctest/test_qset.cpp` to interact with the contract and return the result of update aggregate.

Also we should double check the `test_oracle` tests

File: program/src/oracle/upd_aggregate.h [hunk 1]
```diff
@@ -205,39 +206,76 @@ static inline void upd_aggregate( pc_price_t *ptr, uint64_t slot )
...
+           slot_diff >= 0 && slot_diff <= PC_MAX_SEND_LATENCY ) {
+        vidx[numv++] = i; // FIXME: GRAFT FOR OLD PRICE MODEL
+        prcs[ nprcs++ ] = price - conf; // No overflow as 0 < conf < price
+        prcs[ nprcs++ ] = price;
+        prcs[ nprcs++ ] = price + conf; // No overflow as 0 < conf <= INT64_MAX-price
+      }
+    }
+
+    // too few valid quotes
+    ptr->num_qt_ = numv; // FIXME: TEMPORARY GRAFT (ALL RETURN PATHS SET NUM_QT TO SOMETHING SENSIBLE)
```
Comment (jayantk) [inline program/src/oracle/upd_aggregate.h:237]: why is this temporary? seems like if we delete the old aggregation logic, this should still be set here?

File: program/src/oracle/upd_aggregate.h [hunk 2]
```diff
@@ -205,39 +206,76 @@ static inline void upd_aggregate( pc_price_t *ptr, uint64_t slot )
...
+      return;
+    }
   }
-  qsort_int64( prcs, nprcs );
+
+  // FIXME: MOST OF THE REST OF THE CODE HERE CAN PROBABLY BE REMOVED /
+  // CONDENSED AND/OR MADE MORE EFFICIENT.  AND IT PROBABLY WILL
+  // REPLACED SOON BY THE VWAP MODEL UNDER DEVELOPMENT.  IT IS KEPT HERE
+  // TO KEEP THE CURRENT VWAP CALCULATION AS CLOSE AS POSSIBLE TO WHAT
+  // HOW IT CURRENTLY WORKS.
```
Comment (jayantk) [inline program/src/oracle/upd_aggregate.h:273]: let's actually do this comment and delete the old price model from below. we can keep the old twap computation and feed the new aggregate price / confidence into it. I believe most of the code below is dead.

File: program/src/oracle/test_oracle.c [hunk 3]
```diff
@@ -605,8 +604,8 @@ Test( oracle, upd_aggregate ) {
   px->comp_[px->num_++].latest_ = p4;
   px->comp_[px->num_++].latest_ = p2;
   upd_aggregate( px, 1001 );
-  cr_assert( px->agg_.price_ == 235 );
-  cr_assert( px->agg_.conf_ == 99 );
+  cr_assert( px->agg_.price_ == 245 );
+  cr_assert( px->agg_.conf_ == 85 );
```
Comment (jayantk) [inline program/src/oracle/test_oracle.c:608]: checked these numbers in my head and they seem right

File: program/src/oracle/util/align.h [hunk 4]
```diff
@@ -0,0 +1,57 @@
+#ifndef _pyth_oracle_util_align_h_
+#define _pyth_oracle_util_align_h_
+
+#include "compat_stdint.h" /* For uintptr_t */
+#include <stdalign.h>      /* For alignof */
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
```
Comment (jayantk) [inline program/src/oracle/util/align.h:10]: my brain hurts at this point. i'm assuming these utils are fine. I checked avg carefully. the rest of these seem mechanical and also thoroughly tested.

**Review by jayantk — COMMENTED** [review]
ok i read through this carefully and it largely makes sense. left a couple comments. would really prefer to address anything nontrivial in a separate PR though. will follow up on slack to figure out the best next steps.

File: program/src/oracle/model/test_price_model.c [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    if( memcmp( quote, quote0, sizeof(int64_t)*(size_t)cnt ) ) { printf( "FAIL (sort)\n" ); return 1; }
+
+    uint64_t p25_idx = cnt>>2;
+    uint64_t p50_idx = cnt>>1;
+    uint64_t p75_idx = cnt - (uint64_t)1 - p25_idx;
+    uint64_t is_even = (uint64_t)!(cnt & (uint64_t)1);
+
+    if( val[0]!=quote[ p25_idx ] ) { printf( "FAIL (p25)\n" ); return 1; }
+    if( val[1]!=avg_2_int64( quote[ p50_idx-is_even ], quote[ p50_idx ] ) ) { printf( "FAIL (p50)\n" ); return 1; }
+    if( val[2]!=quote[ p75_idx ] ) { printf( "FAIL (p75)\n" ); return 1; }
```
[the remark this answers is not in the document]
Comment (ali-bahjati) [reply program/src/oracle/model/test_price_model.c:59]: There is another testing suite for testing actual/expected, in `pyth/tests/qset`. It's failing now and we should fix it.

**Review by jayantk — COMMENTED** [review]
LGTM. We decided to put this in a separate branch (so we can hotfix `main` if necessary), so I think you can push this to a branch called `v2` and we can start sending PRs against that.

**Review by ali-bahjati — COMMENTED** [review]
I read through the unused previous logic, I think it is safe to remove them. Let's discuss it.

**Comment by Reisen** [comment]
Pushed as branch `v2`, so no merge required. New aggregation PR's can be made against `v2`.

**Comment by Reisen** [comment]
Pushed as branch `v2`, so no merge required. New aggregation PR's can be made against `v2`.

**Pull request closed without merging**